### DISCUSSION TITLE
How to create the user_created field when connecting an existing database?

### DISCUSSION THREAD

**Discussion by kevincam3** [directus/directus#16547] [Archive]
I had tables that I imported into Directus. One of my tables is named user_created and now I'm trying to figure out how to make it work the like "user_created" field that Directus creates when creating a new collection. 

Is this possible? If not is there a workaround or something I can do manually in the database?

**Comment by Enhed** [comment]
its m2o field in schema:  
<img width="628" alt="image" src="https://user-images.githubusercontent.com/36981278/202964586-7d8c0f59-bac9-4b88-bfec-6e96bc769b32.png">

**Comment by kevincam3** [comment]
Unfortunately my field is a string so I do get that option. My database was an existing database so I didn't create the fields within Directus

![image](https://user-images.githubusercontent.com/870934/202968047-005a88d3-c3a1-4ad0-99f6-1d33f1510a89.png)

**Reply by Enhed** [reply]
In that case you can to make an extension hook like this:  
more about extensions: https://docs.directus.io/extensions/introduction.html
```ts
import { defineHook } from '@directus/extensions-sdk'

export default defineHook(({ action }, { services }) => {
	const nameOfYourCollection = 'books' // for example

	const { ItemsService } = services

	action(`${nameOfYourCollection}.items.create`, async (meta, context) => {
		const { database, schema, accountability } = context

		const service = new ItemsService(nameOfYourCollection, {
			schema: schema!, knex: database
		})

		if(!accountability) return

		await service.updateOne(meta.key, {
			user_id: accountability.user
		})
	})
})

```

**Comment by erondpowell** [comment]
@kevincam3 @Enhed 

~~FYI, you don't need to create a custom hook for this. 
You can add those `optional system fields` at any time. 
Check this thread out: https://github.com/directus/directus/discussions/4686
If that doesn't explain it, I'll make a short video for you.~~

I didn't notice at first your existing user_id field was a string.

**Reply by erondpowell** [reply]
This is a silly question on my end because there's a few obvious problems I can see, but.... is there any reason you can't *create* or *create and switch* to a UUID field now?

Also, fwiw I _think_ you could do this same logic with a flow. 
But tbh, I've been entirely focused in Directus Core. So CC @freekrai

**Reply by kevincam3** [reply]
Eron,

The problem is my database is MySQL and using a UUID as the primary key will kill performance. I think someone should look into this. It's not a problem with Postgres which seems to be the database used by Directus Cloud. So this performance issue could be a problem with someone building the tables via the Directus interface if the default is UUID for the primary key. 

Something to use to undestand this
https://www.youtube.com/watch?v=Y5mWz4vK10A

**Reply by erondpowell** [reply]
This issue is a bit too deep for my current skill/knowledge of DB.
CC @directus/team

**Reply by br41nslug** [reply]
The relation foreignKey to `directus_users` needs to be UUID because the primary key for `directus_users` is *already* of type `UUID` (this goes for most if not all core tables). If the use of `UUID`s is a performance issue that could be a good improvement suggestion.

**Reply by kevincam3** [reply]
Yes, I think this should be changed. You can search all over the internet and you'll find people discussing this performance issue with MySQL using UUID as a primary key.

**Reply by br41nslug** [reply]
There are very good reasons to use UUIDs so it is not as easy as "let's change it because one of the supported vendors has reduced performance with it". As a MySQL specific workaround saving the UUID as its numeric value or string might be an option for when you create your suggestion thread 😄 Looking forward to seeing your suggestions

**Reply by kevincam3** [reply]
I understand but it's also not like we're talking about a small number of users, MySQL is arguably the most used database in the world. I reviewed people suggesting to save it as a number and it's still a problem.

Maybe give users a choice on installation. Is there a way I can change it on my locally installed instance?

**Reply by br41nslug** [reply]
There is no way to change this currently besides making a proper feature request. The original question of this thread has been answered (the foreignKey needs to be a UUID type to work with the existing core tables). Lets continue the further discussion either in a separate feature/improvement request discussion dedicated to this (or even an issue depending on how reproducable the performance you're seeing are).
